bindings/java/src/test/java/org/apache/opendal/OperatorTest.java

```diff
@@ -21,55 +21,175 @@
+import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.Test;
-import org.junit.jupiter.api.io.TempDir;
 
 public class OperatorTest {
-    @TempDir
-    private static Path tempDir;
+
+    private static final String SCHEMA_KEY = "OPENDAL_TEST_SCHEMA";
+
+    protected static Optional<Operator> opOptional;
+
+    protected static Optional<BlockingOperator> blockingOpOptional;
+
+    @BeforeAll
+    public static void init() {
+        String schema = System.getProperty(SCHEMA_KEY);
+        assertTrue(schema != null && schema.length() > 0, () -> SCHEMA_KEY + " is not set");
+
+        opOptional = Utils.init(schema);
+
+        blockingOpOptional = Utils.initBlockingOp(schema);
+    }
+
+    @AfterAll
+    public static void clean() {
+        opOptional.ifPresent(op -> op.close());
+        blockingOpOptional.ifPresent(op -> op.close());
+    }
 
     @Test
-    public void testCreateAndDelete() {
-        Map<String, String> params = new HashMap<>();
-        params.put("root", "/tmp");
-        @Cleanup Operator op = new Operator("Memory", params);
-
-        op.write("testCreateAndDelete", "Odin").join();
-        assertThat(op.read("testCreateAndDelete").join()).isEqualTo("Odin");
-        op.delete("testCreateAndDelete").join();
-        assertThatThrownBy(() -> op.stat("testCreateAndDelete").join())
+    public void testBlockingWrite() {
+        if (!blockingOpOptional.isPresent()) {
+            return;
+        }
+        BlockingOperator blockingOp = blockingOpOptional.get();
+        Capability cap = blockingOp.info().fullCapability;
+        if (!cap.write || !cap.read) {
+            return;
+        }
+
+        String path = UUID.randomUUID().toString();
+        byte[] content = Utils.generateBytes();
+        blockingOp.write(path, content);
+
+        Metadata metadata = blockingOp.stat(path);
+
+        assertEquals(content.length, metadata.getContentLength());
+
+        blockingOp.delete(path);
+        assertThatThrownBy(() -> blockingOp.stat(path))
                 .is(OpenDALExceptionCondition.ofAsync(OpenDALException.Code.NotFound));
     }
 
     @Test
-    public void testAppendManyTimes() {
-        Map<String, String> params = new HashMap<>();
-        params.put("root", tempDir.toString());
-        @Cleanup Operator op = new Operator("fs", params);
+    public void testBlockingRead() {
+        if (!blockingOpOptional.isPresent()) {
+            return;
+        }
+        BlockingOperator blockingOp = blockingOpOptional.get();
+        Capability cap = blockingOp.info().fullCapability;
+        if (!cap.write || !cap.read) {
+            return;
+        }
+
+        Metadata metadata = blockingOp.stat("");
+        assertTrue(!metadata.isFile());
+
+        String path = UUID.randomUUID().toString();
+        assertThatThrownBy(() -> blockingOp.stat(path))
+                .is(OpenDALExceptionCondition.ofAsync(OpenDALException.Code.NotFound));
+
+        String content = Utils.generateString();
```

Comment:
content should be bytes instead of string.

Reply:
Since the `read` method returns string, it is unreadable when writing bytes.

<details>
<summary>error</summary>

```plaintext
java.util.concurrent.CompletionException: org.apache.opendal.OpenDALException: Unexpected (permanent) at  => invalid utf-8 sequence of 1 bytes from index 0, source: invalid utf-8 sequence of 1 bytes from index 0
        at java.base/java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:413)
        at java.base/java.util.concurrent.CompletableFuture.join(CompletableFuture.java:2118)
        at org.apache.opendal.DemoTest.testRead(DemoTest.java:33)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
        at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
        at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
        at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
        at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
        at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
        at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
        at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
        at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
        at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
        at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
        at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
        at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
        at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
        at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
        at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:50)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:184)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:148)
        at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:122)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
        at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: org.apache.opendal.OpenDALException: Unexpected (permanent) at  => invalid utf-8 sequence of 1 bytes from index 0, source: invalid utf-8 sequence of 1 bytes from index 0
```
</details>

Reply:
> Since the `read` method returns string

It's a bug that `read` returns string, we should fix it first.

Reply:
A fix is proposed at https://github.com/apache/incubator-opendal/pull/3153.